File: kernel/arch/x86_64/system/acpi/hpet.c
// High Precision Event Timer table

#include <menix/memory/pm.h>
#include <menix/system/acpi/acpi.h>
#include <menix/system/time/clock.h>
#include <menix/util/log.h>

#include <hpet.h>
#include <uacpi/acpi.h>
#include <uacpi/tables.h>

typedef struct
{
	volatile HpetRegisters* regs;
	u32 period;
	ClockSource cs;
} Hpet;

static usize hpet_get_elapsed_ns();
static void hpet_reset();

static Hpet hpet = {
	.cs = {.name = "hpet", .get_elapsed_ns = hpet_get_elapsed_ns, .reset = hpet_reset},
};

static usize hpet_get_elapsed_ns()
{
	// Convert femtoseconds to nanoseconds.
	return hpet.regs->main_counter * (hpet.period / 1000000);
}

static void hpet_reset()
{
	hpet.regs->main_counter = 0;
}

static void hpet_setup(PhysAddr addr)
{
	print_log("acpi: HPET at 0x%p\n", addr);
	hpet.regs = pm_get_phys_base() + addr;

	// Get the period.
	hpet.period = (u32)(hpet.regs->capabilities >> 32);

	// Enable timer.
	hpet.regs->configuration |= 1;

	print_log("acpi: Enabled HPET.\n");

	clock_register(&hpet.cs);
}

void hpet_init()
{
	uacpi_table hpet_table;
	if (uacpi_table_find_by_signature("HPET", &hpet_table) != UACPI_STATUS_OK)
	{
		print_error("acpi: Failed to get the HPET table!\n");
		return;
	}

	struct acpi_hpet* hpet = hpet_table.ptr;

	hpet_setup(hpet->address.address);
	uacpi_table_unref(&hpet_table);
}

File: kernel/src/arch/x86_64/platform/hpet.rs
use crate::generic::{
    clock::{ClockError, ClockSource},
    memory::mmio::Mmio,
};
use uacpi_sys::{
    UACPI_STATUS_OK, acpi_hpet, uacpi_table, uacpi_table_find_by_signature, uacpi_table_unref,
};

pub struct Hpet {
    regs: Mmio,
    period: u32,
}

mod regs {
    use crate::generic::memory::mmio::Register;

    pub const CAPABILITIES: Register<u64> = Register::new(0);
    pub const CONFIGURATION: Register<u64> = Register::new(0x10);
    pub const MAIN_COUNTER: Register<u64> = Register::new(0xF0);
}

impl ClockSource for Hpet {
    fn name(&self) -> &'static str {
        "hpet"
    }

    fn get_priority(&self) -> u8 {
        75
    }

    fn reset(&mut self) {
        self.regs.write(regs::MAIN_COUNTER, 0 as u64);
    }

    fn get_elapsed_ns(&self) -> usize {
        return (self.regs.read(regs::MAIN_COUNTER) * self.period as u64 / 1_000_000) as usize
            as usize;
    }
}

impl Hpet {
    pub fn new() -> Result<Self, ClockError> {
        let mut table = uacpi_table::default();
        let uacpi_status =
            unsafe { uacpi_table_find_by_signature(c"HPET".as_ptr(), &raw mut table) };

        if uacpi_status != UACPI_STATUS_OK {
            return Err(ClockError::Unavailable);
        }

        let hpet: *mut acpi_hpet = unsafe { table.__bindgen_anon_1.ptr } as *mut acpi_hpet;
        let mut mmio = unsafe { Mmio::new_mmio(((*hpet).address.address as usize).into(), 0x1000) };
        unsafe { uacpi_table_unref(&raw mut table) };

        // Enable timer.
        mmio.write(regs::CONFIGURATION, mmio.read(regs::CONFIGURATION) | 1);
        let period = (mmio.read(regs::CAPABILITIES) >> 32) as u32;

        return Ok(Hpet {
            regs: mmio,
            period: period,
        });
    }
}
